### ISSUE TITLE
Proposals for naming and package structure

### ISSUE THREAD

**Issue by NthPortal** [scala/scala-library-next#24]
Proposals:

1. Only names that are planned to exist in a future Scala release can be outside of a `next` (sub-)package. Names that will not exist in a future Scala release MUST NOT be outside of a `next` (sub-)package.
   a. Corollary: Extension methods MUST go in `next` (sub-)packages.
   b. Possible extension: All names that are planned to exist in a future Scala release MUST be outside of `next` (sub-)packages.
2. Extension method value classes SHOULD be named `Next<Name>Extensions`, where `<Name>` is the name of the type.
   a. For example, extension methods for `LazyList` would go in value class named `NextLazyListExtensions`.
   b. Extension methods for an object that is a companion SHOULD be named `Next<Name>CompanionExtensions`.

Something something RFC 2119

**Comment by sjrd** [comment]
> Names that will not exist in a future Scala release MUST NOT be outside of a next (sub-)package.

The triple negation was very hard for me too read, and I understood it the wrong way the first two times I read the sentence. Perhaps consider:
```diff
-Names that will not exist in a future Scala release MUST NOT be outside of a next (sub-)package.
+Names that will not exist in a future Scala release MUST be inside of a next (sub-)package.
```

**Comment by NthPortal** [comment]
I don't know why that slipped by me. yes, fixed

**Comment by BalmungSan** [comment]
@NthPortal How should the object holding the value class be named? Just `<Name>Extensions`?
So for example, if I want some additional methods on **LazyList** I would `import scala.collection.immutable.next.LazyListExtensions._`?

**Comment by NthPortal** [comment]
IMO, it can go directly in the package object.

**Comment by BalmungSan** [comment]
@NthPortal so importing `next._` would import all extensions?

**Comment by NthPortal** [comment]
correct. and if you wanted to import just one, you'd import `next.NextLazyListExtensions` or whatever

**Comment by martijnhoekstra** [comment]
Having to import next.something would mean that cross-building 3.0/3.1 won't be possible unless we ship this with empty next instances for 3.1. that'd be lame.

**Comment by NthPortal** [comment]
then just import `next._`

**Comment by BalmungSan** [comment]
So there will be many next` objects; for example, `scala.collection.next`, `scala.collection.immutable.next` and `scala.collection.mutable.next`.

The problem would be how to organize the code because then all extensions would be in a single big file.
_(Also, using explicit names of implicits is discouraged)_
I propose to follow a structure similar to the one in **cats**, something like this:

```scala
// file: src/main/scala/scala/next/collection/Next<Name>Extensions.scala
package scala.next
package collection

trait Next<Name>Extensions {
  implicit final def scalaNextSyntaxFor<Name>[A](coll: <Name>[A]): Next<Name>Ops[A] =
    new Next<Name>Ops(col)
}

private[next] final class Next<Name>Ops[A](private val col: <Name>[A]) extends AnyVal {
  // Extension methods go here.
}
```

```scala
// file: src/main/scala/scala/next/collection/All.scala
package scala.next.collection

trait NextCollectionAllExtensions extends Next<Name>Extensions with ...
```

```scala
// file: src/main/scala/scala/next/collection/package.scala

package scala.next

package object collection {
  object all extends NextCollectionAllExtensions
  object <name> extends Next<Name>Extensions
}
```

```scala
// file: src/main/scala/scala/next/package.scala

package scala

package object next {
  object implicits extends collection.NextCollectionAllExtensions with ...
}
```

Not sure how this would affect cross-building.

**Comment by SethTisue** [comment]
@som-snytt would `-Yimports` be helpful here? as a way of allowing the imports to be different in different Scala versions, as needed, without the user having to alter their actual source code

**Comment by martijnhoekstra** [comment]
Then you'd still need to ship a version with an empty next package for 3.1. which means inflicting the transitive dependency on your dependents.

That can be sort of worked around with people taking a provided dependency for 3.1.

**Comment by sjrd** [comment]
> Having to import next.something would mean that cross-building 3.0/3.1 won't be possible unless we ship this with empty next instances for 3.1. that'd be lame.

I think that shipping empty `next` instances is exactly what we'll do.

Anyway, they probably won't stay empty as we then start adding new stuff coming in 3.2 or later.

**Comment by martijnhoekstra** [comment]
Julien suggested on https://github.com/scala/scala-library-next/issues/25 that there wouldn't be 3.2 stuff in this library.

Per version `-Yimports` is probably at least as invasive/cumbersome for users as having the lame dependency on the empty next package and declaring it provided for 3.1.

**Comment by NthPortal** [comment]
I've noticed we still have a shadowing problem for names which appear in multiple packages (e.g. `Set`, `Seq`), so I've updated the proposals to account for that

**Comment by BalmungSan** [comment]
Did you come to a conclusion about this?

So I could update #23 and open a follow up for `groupByTo` & `groupMapTo`

**Comment by NthPortal** [comment]
Votes (see the following comment for more granular votes)

In favour of (👍) or against (👎) proposal 1 **_in its entirety_**

In favour of (🚀) or agains (👀) proposal 2 **_in its entirety_**

**Comment by NthPortal** [comment]
Granular votes (by reacting):

In favour of proposals 1 and 1a but against 1b (👀)
In favour of proposals 2 and 2a but against 2b (👎)
In favour of proposals 2 and 2a but against 2c (😕)

**Comment by SethTisue** [comment]
This all looks good, except I don't know what I think about 1b. What are the pros and cons there?

**Comment by NthPortal** [comment]
I'll give an example of something that 1b affects.

Currently, `scala.collection.concurrent.Map` does not have a companion, unlike other `Map` types (see scala/bug#12115). We cannot add it in that package in scala-library-next, because the companion needs to be in the same compilation unit. If we were to add such a companion, it would have to be an `object Map` in `package scala.collection.concurrent.next`; however, that would violate 1b.

If we were to reject 1b and add `s.c.c.next.Map`, we would need an alias `val Map = s.c.c.Map` in the Scala 3.1 version of this library, and couldn't just have an empty package.

**Comment by BalmungSan** [comment]
@NthPortal as far as I can tell the current proposals only take into consideration the naming of the **implicit value classes**, but does not define where they should be defined.

As I said before, I do not t think that having all of them in the `package object` to be a good idea, the file will be pretty big and searching for the implementation of a particular method would be difficult. So if that will be the way to go, then it should be at least clear in the proposals.

Note that if the underlying motivation is to provide a single `next._` import, then you can still split the implementations into a `final class` with the **extension method** and a `trait` with the **implicit conversion** and then use a regular `object` that extends all the `traits`.

------

> Only names that are planned to exist in a future Scala release can be outside of a next (sub-)package. Names that will not exist in a future Scala release MUST be in a next (sub-)package.

AFAIK the idea is that if something will not exist in a future scala version _(or it is still not clear if it should be part or not of)_ then it shouldn't be here in the first place. Thus, this proposal wouldn't make sense, right?

> a. Corollary: Extension methods MUST go in next (sub-)packages.

**Comment by NthPortal** [comment]
> the current proposals only take into consideration the naming of the **implicit value classes**, but [do] not define where they should be defined.

These proposals also make some constraints on what MUST go in a `next` (sub-)package, and perhaps what MUST NOT as well.

> As I said before, I do not [...] think that having all of them in the `package object` to be a good idea, the file will be pretty big and searching for the implementation of a particular method would be difficult.

By all means feel free to create a proposal that specifies that. These proposals were not intended to specify everything that needs to be decided about the library's structure.

It's worth noting that value classes MUST live inside (top-level) objects, so having them split up necessitates a separate wrapper object for each extension method value class.

--------

> > Only names that are planned to exist in a future Scala release can be outside of a next (sub-)package. Names that will not exist in a future Scala release MUST be in a next (sub-)package.
>
> AFAIK the idea is that if something will not exist in a future scala version _(or it is still not clear if it should be part or not of)_ then it shouldn't be here in the first place.

I was very deliberate with my wording. _Names_ that will not exist in a future Scala version have to be in a `next` (sub-)package. For example, since 3.1 will not have the name `NextLazyListExtensions`, that value class/extension MUST be in a `next` (sub-)package. I said "names", not "things".

The reason 1a is a corollary is because, at least using Scala 2 syntax, extension methods _require_ a value class and an implicit conversion to it (which are usually combined as in an implicit value class), both of which are names that are not intended to exist in a future Scala version. Thus, they MUST be in `next` (sub-)packages.

> > a. Corollary: Extension methods MUST go in next (sub-)packages.
>
> I would also add this to the second proposal.

Proposal 1 is about package structure; proposal 2 is about the naming of value classes for extension methods. That's why 1a is where it is. Additionally, 1 strictly implies 1a; they are inseparable.

Also, I'm not sure why a particular item should be proposed twice.

**Comment by martijnhoekstra** [comment]
I would really prefer to be able to import a name that does not require shipping an empty package which forces that empty package onto dependents as a transitive dependency without jumping through the error-prone hoops of making it provided/not provided per scala version.

Another solution to that problem could be for the next stdlib to provide an empty next package. I would find that a bit less onerous than shipping transitively dependent (and possibly version-conflicting) empty jars.

As for proposal 2, it depends a bit on what will happen after 3.1. If this library is then retired until scala 4 as suggested by Julien, the scheme sounds good. If new, scala 3.2 additions will be added to this library as suggested by seb, and it's still intended to be used in 3.0, will the naming scheme then still work, or will there be clashes?

As long as we won't have to end up with the moral equivalent of `import realnextsecond.NexNextLazyListExtensions` in 3.1-next because we don't want to harp on too much about what's not decided yet anyway, I'm happy with the scheme as proposed.

**Comment by BalmungSan** [comment]
> These proposals also make some constraints on what MUST go in a next (sub-)package, and perhaps what MUST NOT as well.

Yeah, I mean more about the structure of a change, which files to create / edit and what should be on each of those.

> By all means feel free to create a proposal that specifies that.

Isn't that what I did [here](https://github.com/scala/scala-library-next/issues/24#issuecomment-716770542) and that I repeated here: _"then you can still split the implementations into a final class with the extension method and a trait with the implicit conversion and then use a regular object that extends all the traits"_.

> I said "names", not "things".

Makes sense, sorry for the confusion.

> Proposal 1 is about package structure; proposal 2 is about the naming of value classes for extension methods.
> Also, I'm not sure why a particular item should be proposed twice.

Ah, I thought they were two options, not complementary _(like radio buttons instead of checkboxes)_ it makes sense now.

**Comment by BalmungSan** [comment]
Thinking about it and given that this project doesn't want to provide a granular set of imports but rather just a single `next._` _(or at least one `next` per package)_. We do not need something as complex as my previous proposal.

But rather just something like this:

```scala
// file: src/main/scala/scala/next/Next<Name>Extensions.scala

package scala.next

private[next] final class Next<Name>Ops[A](private val col: <Name>[A]) extends AnyVal {
  // Extension methods go here.
}
```

```scala
// file: src/main/scala/scala/next/package.scala

package scala

package object next {
  implicit final def scalaNextSyntaxFor<Name>[A](coll: <Name>[A]): Next<Name>Ops[A] =
    new Next<Name>Ops(col)
}
```

And maybe this is what you _(@NthPortal)_ mean when you said:

> IMO, it can go directly in the package object.

----

One open discussion for this proposal would be:

1. One single `next` package for all extensions: `scala.next._`
2. Multiple `next` packages for each package of the stdlib: `scala.collection.next._`, `scala.collection.immutable.next._`, _etc_.

-----

Vote with:
In favour of this proposal going with option 1. (👍)
In favour of this proposal going with option 2. (🚀) 
Completely against this proposal. (👎)

**Comment by NthPortal** [comment]
> Isn't that what I did here and that I repeated here

I meant a separate issue/ticket

> ```scala
> // file: src/main/scala/scala/next/Next<Name>Extensions.scala
> 
> package scala.next
>
> private[next] final class Next<Name>Ops[A](private val col: <Name>[A]) extends AnyVal {
>   // Extension methods go here.
> }
> ```

Scala 2 does not permit top-level value classes

**Comment by BalmungSan** [comment]
> I meant a separate issue/ticket

Isn't this issue for all the proposals and choosing one?

> Scala 2 does not permit top-level value classes

Ham yes it does, you are confusing it with **implicit classes**, that is why the **implicit conversion** is in the `package object`. This is a simplified version of what libraries like **cats** or [**neotypes**](https://github.com/neotypes/neotypes/tree/master/core/src/main/scala/neotypes/implicits/syntax) do.

**Comment by NthPortal** [comment]
> yes it does, you are confusing it with implicit classes

you're right, my mistake

> Isn't this issue for all the proposals and choosing one?

no. it's an issue for the 2 proposals (and their sub-proposals/extensions) listed in the issue description. also, again, the proposals in this issue's description are not options, one of which gets chosen. they are two independent proposals, each to be accepted or rejected individually.

**Comment by BalmungSan** [comment]
> I meant a separate issue/ticket

Done #46 

> no. it's an issue for the 2 proposals (and their sub-proposals/extensions) listed in the issue description. also, again, the proposals in this issue's description are not options, one of which gets chosen. they are two independent proposals, each to be accepted or rejected individually.

Yeah yeah, my point was that my proposal was also complementary with yours, so I thought it made sense to be in the same issue, but ok.

I also deleted all my comments in this issue that I believe are no longer relevant.

**Comment by NthPortal** [comment]
follow-up at #46

**Comment by som-snytt** [comment]
Was there already a vote on branch names?

**Comment by NthPortal** [comment]
I propose that we don't wait on finalizing this and #46 to merge PRs, and we can just refactor the library once we've decided on those.
